### PR TITLE
Feature: Apache Drill query runner

### PR DESCRIPTION
**Pull request by dharmicdata**
New query runner for connecting to Apache drill.
Added drill as a query runner in settings.py.
Added drill_setup.sh in in the setup/ubuntu folder. It is not run automatically in the bootstrap script.

### CODE REVIEW

File: redash/query_runner/drill.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+try:
+    import pyodbc
+    enabled = True
+    logger.warning("LD_LIBRARY_PATH: {}".format(os.environ.get('LD_LIBRARY_PATH')))
+    logger.warning("ODBCINI: {}".format(os.environ.get('ODBCINI')))
+    logger.warning("MAPRDRILLINI: {}".format(os.environ.get('MAPRDRILLINI')))
+except ImportError, e:
+    logger.warning("Missing dependencies. Please install pyodbc.")
+    logger.warning("You can use pip: pip install pyodbc")
```
Comment (arikfr) [inline redash/query_runner/drill.py]: We no longer use these messages. Please remove.

File: redash/query_runner/drill.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+from redash.utils import JSONEncoder
+
+logger = logging.getLogger(__name__)
+
+try:
+    import pyodbc
+    enabled = True
+    logger.warning("LD_LIBRARY_PATH: {}".format(os.environ.get('LD_LIBRARY_PATH')))
+    logger.warning("ODBCINI: {}".format(os.environ.get('ODBCINI')))
+    logger.warning("MAPRDRILLINI: {}".format(os.environ.get('MAPRDRILLINI')))
```
Comment (arikfr) [inline redash/query_runner/drill.py]: Please remove these warnings or move them to `debug` level.

File: redash/query_runner/drill.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+
+class Drill(BaseQueryRunner):
+    @classmethod
+    def configuration_schema(cls):
+        return {
+            "type": "object",
+            "properties": {
+                "ConnectionType": {
```
Comment (arikfr) [inline redash/query_runner/drill.py]: Please use consistent naming for the configuration properties. Preferably snake cases, i.e.: `connection_type`.

Also provide a `title` to give a friendly name, specially for the ZK ones.

File: redash/query_runner/drill.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+                for table_name in map(lambda a: a['name'], self._run_query_internal(query_to_run % schema_name)):
+                    columns = map(lambda a: a['friendly_name'], self._run_query_internal_columns(columns_query % (schema_name, table_name)))
+
+                    if schema_name != 'INFORMATION_SCHEMA' and schema_name != 'sys':
+                        table_name = '{}.`{}`'.format(schema_name, table_name)
+
+                    schema[table_name] = {'name': table_name, 'columns': columns}
+        except Exception, e:
+            raise sys.exc_info()[1], None, sys.exc_info()[2]
```
Comment (arikfr) [inline redash/query_runner/drill.py]: If you don't handle any exception and just re-raise it, why have the try/except at all?

**Comment by arikfr** [comment]
Thanks! 

I've added a few comments, mainly on style. Later on I will setup Drill to give it a proper test so might have some more comments.

**Comment by breath103** [comment]
👍 love to see get merged. any progress? i'm running re:dash on production AND apache drill on production too so let me know if need help on testing

**Comment by arikfr** [comment]
@breath103 if you can test it on your setup and give feedback on if you had issues (as I never had the chance to test it myself :-(). There are a few style fixes to do, but I can do them myself after merging, once we know it's OK.

**Comment by breath103** [comment]
@arikfr ok let me try out
@dharmicdata by the way, how do you think about implementing this with Apache Drill HTTP API? i feel like using JDBC / ODBC seems making this pretty much complicated, and since queries on apache drill usually take few minutes to be solved, that would be no big difference i feel like

**Comment by arikfr** [comment]
@breath103 using the HTTP API is an interesting idea, and will simplify this greatly. Do you know if there are any options that we lose if we switch to it?

**Comment by breath103** [comment]
well i can't really think of any... yet. basically Apache Drill supports ANSI standard for SQL, so even it's REST API is about passing SQL through the HTTP and getting result with JSON.
it also supports command like "SHOW SCHEMAS;", "SHOW TABLES".. etc so i guess it would work. even though that means we force people to open the extra port (8047 usually) but not sure that's the major problem here 

+ the format of Apache Drill's REST API returns is same format with the JSON format the Re:dash URL datasource supports, {columns: [], rows: []},
so basically, we could even use Apache Drill if the URL datasource supports POST call and Body.

**Comment by arikfr** [comment]
@breath103 it's very similar, but the columns object is different. From the example in their docs, all the values returned as strings. Is it really the case? Is there some support for authentication?

**Comment by breath103** [comment]
oh i see. types.. that could be tricky i guess. as of authentication, i don't think there is one. setting IP whitelist from the apache drill side would be the only way

**Comment by jeffwidman** [comment]
@dharmicdata Any updates here?

**Comment by arikfr** [comment]
Closing for inactivity :(

**Comment by fcastillao** [comment]
nobody tested this?
